File: Diagnostic2/CodeFixProvider.vb
Imports Microsoft.CodeAnalysis.Rename

<ExportCodeFixProvider(DiagnosticAnalyzer.DiagnosticId, LanguageNames.VisualBasic)>
Friend Class CodeFixProvider
  Implements ICodeFixProvider

  Public Function GetFixableDiagnosticIds() As IEnumerable(Of String) Implements ICodeFixProvider.GetFixableDiagnosticIds
    Return {DiagnosticAnalyzer.DiagnosticId}
  End Function

  Public Async Function GetFixesAsync(document As Document, span As TextSpan, diagnostics As IEnumerable(Of Diagnostic), cancellationToken As CancellationToken) As Task(Of IEnumerable(Of CodeAction)) Implements ICodeFixProvider.GetFixesAsync
    Dim root = Await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(False)

    ' TODO: Replace the following code with your own analysis, generating a CodeAction for each fix to suggest

    Dim diagnosticSpan = diagnostics.First().Location.SourceSpan

    ' Find the type statement identified by the diagnostic.
    Dim declaration = root.FindToken(diagnosticSpan.Start).Parent.AncestorsAndSelf().OfType(Of TypeStatementSyntax)().First()

    ' Return a code action that will invoke the fix.
    Return {CodeAction.Create("Make uppercase", Function(c) MakeUppercaseAsync(document, declaration, c))}
  End Function

  Private Async Function MakeUppercaseAsync(document As Document, typeStmt As TypeStatementSyntax, cancellationToken As CancellationToken) As Task(Of Solution)
    ' Compute new uppercase name.
    Dim identifierToken = typeStmt.Identifier
    Dim newName = identifierToken.Text.ToUpperInvariant()

    ' Get the symbol representing the type to be renamed.
    Dim semanticModel = Await document.GetSemanticModelAsync(cancellationToken)
    Dim typeSymbol = semanticModel.GetDeclaredSymbol(typeStmt, cancellationToken)

    ' Produce a new solution that has all references to that type renamed, including the declaration.
    Dim originalSolution = document.Project.Solution
    Dim optionSet = originalSolution.Workspace.GetOptions()
    Dim newSolution = Await Renamer.RenameSymbolAsync(document.Project.Solution, typeSymbol, newName, optionSet, cancellationToken).ConfigureAwait(False)

    ' Return the new solution with the now-uppercase type name.
    Return newSolution
  End Function
End Class

File: VS2013/StringFormatDiagnostic/Diagnostic1/CodeFixProvider.cs
using System.Collections.Generic;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.CodeFixes;
using Microsoft.CodeAnalysis.CodeActions;
using Microsoft.CodeAnalysis.CSharp;
using Microsoft.CodeAnalysis.CSharp.Syntax;
using Microsoft.CodeAnalysis.Rename;
using Microsoft.CodeAnalysis.Text;

namespace Diagnostic1
{
    [ExportCodeFixProvider(DiagnosticAnalyzer.DiagnosticId, LanguageNames.CSharp)]
    internal class CodeFixProvider : ICodeFixProvider
    {
        public IEnumerable<string> GetFixableDiagnosticIds ( )
        {
            return new [ ] { DiagnosticAnalyzer.DiagnosticId };
        }

        public async Task<IEnumerable<CodeAction>> GetFixesAsync ( Document document, TextSpan span, IEnumerable<Diagnostic> diagnostics, CancellationToken cancellationToken )
        {
            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);

            // TODO: Replace the following code with your own analysis, generating a CodeAction for each fix to suggest

            var diagnosticSpan = diagnostics.First().Location.SourceSpan;

            // Find the type declaration identified by the diagnostic.
            var declaration = root.FindToken(diagnosticSpan.Start).Parent.AncestorsAndSelf().OfType<TypeDeclarationSyntax>().First();

            // Return a code action that will invoke the fix.
            return new [ ] { CodeAction.Create( "Make uppercase", c => MakeUppercaseAsync( document, declaration, c ) ) };
        }

        private async Task<Solution> MakeUppercaseAsync ( Document document, TypeDeclarationSyntax typeDecl, CancellationToken cancellationToken )
        {
            // Compute new uppercase name.
            var identifierToken = typeDecl.Identifier;
            var newName = identifierToken.Text.ToUpperInvariant();

            // Get the symbol representing the type to be renamed.
            var semanticModel = await document.GetSemanticModelAsync(cancellationToken);
            var typeSymbol = semanticModel.GetDeclaredSymbol(typeDecl, cancellationToken);

            // Produce a new solution that has all references to that type renamed, including the declaration.
            var originalSolution = document.Project.Solution;
            var optionSet = originalSolution.Workspace.GetOptions();
            var newSolution = await Renamer.RenameSymbolAsync(document.Project.Solution, typeSymbol, newName, optionSet, cancellationToken).ConfigureAwait(false);

            // Return the new solution with the now-uppercase type name.
            return newSolution;
        }
    }
}